pkg/oam/util/helper.go

```diff
@@ -204,29 +224,72 @@ func FetchWorkloadDefinition(ctx context.Context, r client.Reader, dm discoverym
 	return workloadDefinition, nil
 }
 
-// GenNamespacedDefinitionName generate definition name with customized namespace
-func GenNamespacedDefinitionName(dn string) types.NamespacedName {
-	if dns := os.Getenv(DefinitionNamespaceEnv); dns != "" {
-		return types.NamespacedName{Name: dn, Namespace: dns}
+// GenNamespaceWithCtx gen sysNs and appNs from ctx
+func GenNamespaceWithCtx(ctx context.Context) (string, string) {
+	var sysNs, appNs string
+	if sys := ctx.Value(sysDefinitionNamespace); sys == nil {
+		sysNs = "vela-system"
+	} else {
+		sysNs = sys.(string)
 	}
-	return types.NamespacedName{Name: dn}
+	if app := ctx.Value(appDefinitionNamespace); app == nil {
+		appNs = "vela-system"
+	} else {
+		appNs = app.(string)
+	}
+	return sysNs, appNs
+}
+
+// SetNnamespaceInCtx set system-level namespace and app namespace in context
+func SetNnamespaceInCtx(ctx context.Context, sysNs string, appNs string) context.Context {
+	ctx = context.WithValue(ctx, sysDefinitionNamespace, sysNs)
+	ctx = context.WithValue(ctx, appDefinitionNamespace, appNs)
+	return ctx
 }
 
 // GetWorkloadDefinition  Get WorkloadDefinition
 func GetWorkloadDefinition(ctx context.Context, cli client.Reader,
```

Comment:
why is there no test change related to this?

Reply:
Will add soon.